### PR TITLE
Add default cleanup state machine for VM transformation

### PR DESCRIPTION
**Pull request by fdupont-redhat**
This PR provides the default state machine for VM transformation cleanup. The state machine is triggered from an automate request, as designed in https://github.com/ManageIQ/manageiq-content/pull/357. The request creation passes a variable in `$evm.root`: `service_template_transformation_plan_task_id`.

The process is as follow:

1. Assess the transformation cleanup. It basically sets the required state vars.
2. Kill virt-v2v. The PID of the process and the transformation host information is available in the task. It simply needs a remote command execution to kill the process. `virt-v2v-wrapper` handles cleaning up the destination provider when the virt-v2v command fails, so nothing to do on ManageIQ side.
3. Power on the VM if it was powered on when the migration started.

Associated RHBZ: https://bugzilla.redhat.com/show_bug.cgi?id=1599997

### CODE REVIEW

**Comment by fdupont-redhat** [comment]
@miq-bot add-label transformation
@miq-bot add-label enhancement
@miq-bot add-label gaprindashvili/yes
@miq-bot add-label blocker

File: content/automate/ManageIQ/Transformation/Common.class/__methods__/assesstransformationcleanup.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+          def initialize(handle = $evm)
+            @handle = handle
+          end
+
+          def main
+            task = @handle.vmdb(:service_template_transformation_plan_task).find_by(:id => @handle.root['service_template_transformation_plan_task_id'])
+            raise 'No task found. Exiting' if task.nil?
+            @handle.log(:info, "Task: #{task.inspect}") if @debug
+
+            destination_ems = @handle.vmdb(:ext_management_system).find_by(:id => task.get_option(:destination_ems_id))
```
Comment (mkanoor) [inline content/automate/ManageIQ/Transformation/Common.class/__methods__/assesstransformationcleanup.rb]: @fdupont-redhat 
Can you check if the destination_ems is present before setting the state var.

File: content/automate/ManageIQ/Transformation/Infrastructure/VM/Common.class/__methods__/poweron.rb [hunk 2]
```diff
@@ -10,11 +10,14 @@ def initialize(handle = $evm)
               end
 
               def main
-                task = @handle.root['service_template_transformation_plan_task']
-                if task.get_option(:source_vm_power_state) == 'on'
-                  destination_vm = @handle.vmdb(:vm).find_by(:id => task.get_option(:destination_vm_id))
-                  destination_vm.start
+                if @handle.root['service_template_transformation_plan_task'].blank?
+                  task = @handle.vmdb(:service_template_transformation_plan_task).find_by(:id => @handle.root['service_template_transformation_plan_task_id'])
```
Comment (mkanoor) [inline content/automate/ManageIQ/Transformation/Infrastructure/VM/Common.class/__methods__/poweron.rb]: @fdupont-redhat Check for task being nil

File: content/automate/ManageIQ/Transformation/TransformationHosts/Common.class/__methods__/killvirtv2v.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+            def main
+              require 'json'
+
+              task = @handle.vmdb(:service_template_transformation_plan_task).find_by(:id => @handle.root['service_template_transformation_plan_task_id'])
+
+              # Retrieve transformation host
+              transformation_host = @handle.vmdb(:host).find_by(:id => task.get_option(:transformation_host_id))
+
+              # Skip if virtv2v has not started yet or is already finished
```
Comment (mkanoor) [inline content/automate/ManageIQ/Transformation/TransformationHosts/Common.class/__methods__/killvirtv2v.rb]: @fdupont-redhat 
Can all of these skips be in a separate function called validate or valid and you call the remote command only if its valid.

File: content/automate/ManageIQ/Transformation/Infrastructure/VM/Common.class/__methods__/poweron.rb (hunk 2 above)
Comment (fdupont-redhat) [reply content/automate/ManageIQ/Transformation/Infrastructure/VM/Common.class/__methods__/poweron.rb]: I check if task is nil, as well as VM.

File: content/automate/ManageIQ/Transformation/Common.class/__methods__/assesstransformationcleanup.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      module Common
+        class AssessTransformationCleanup
+          def initialize(handle = $evm)
+            @handle = handle
+          end
+
+          def main
+            task = @handle.vmdb(:service_template_transformation_plan_task).find_by(:id => @handle.root['service_template_transformation_plan_task_id'])
+            raise 'No task found. Exiting' if task.nil?
+            @handle.log(:info, "Task: #{task.inspect}") if @debug
```
Comment (gmcculloug) [inline content/automate/ManageIQ/Transformation/Common.class/__methods__/assesstransformationcleanup.rb]: `@debug` is referenced here but never defined.

File: content/automate/ManageIQ/Transformation/TransformationHosts/Common.class/__methods__/killvirtv2v.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
+module ManageIQ
+  module Automate
+    module Transformation
+      module TransformationHost
+        module Common
+          class VMCheckTransformed
+            def initialize(handle = $evm)
+              @debug = true
```
Comment (gmcculloug) [inline content/automate/ManageIQ/Transformation/TransformationHosts/Common.class/__methods__/killvirtv2v.rb]: and `@debug` is referenced here but never used.

File: content/automate/ManageIQ/Transformation/Common.class/__methods__/assesstransformationcleanup.rb (hunk 4 above)
Comment (fdupont-redhat) [reply content/automate/ManageIQ/Transformation/Common.class/__methods__/assesstransformationcleanup.rb]: Fixed with default value to `false`.

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit a1db90493bec152af05c2ffbfed122dd46d3d385
Author: Greg McCullough <gmccullo@redhat.com>
Date:   Mon Jul 30 18:45:34 2018 -0400

    Merge pull request #378 from fdupont-redhat/v2v_cleanup_add_default_state_machine
    
    Add default cleanup state machine for VM transformation
    (cherry picked from commit fe3fe985101d38eee66c050562d3a7de8666e6ea)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1608758
```

**Pull request merged**